### ISSUE TITLE
Deallocating large BTreeMap takes very long time

### ISSUE THREAD

**Issue by nagisakuya** [rust-lang/rust#126145]
When deallocating large BTreeMap takes very long time. Deallocating a BTreeMap with 15GB size with 60M elements(each element is 250 byte) takes around 20 seconds on my computer. 
In my plogram, when code is exiting, that deallocating occures and takes long time. But in that case, deallocating is not needed because just freeing all allocated memory pages by os is enough.

**Comment by veera-sivarajan** [comment]
Hi, thanks for opening this issue. It would be helpful if you could attach some code to reproduce the issue.

@rustbot label -needs-triage +S-needs-repro +I-slow

**Comment by workingjubilee** [comment]
> When deallocating large BTreeMap takes very long time. Deallocating a BTreeMap with 15GB size with 60M elements(each element is 250 byte) takes around 20 seconds on my computer. In my plogram, when code is exiting, that deallocating occures and takes long time. But in that case, deallocating is not needed because just freeing all allocated memory pages by os is enough.

Hmm. If you know it is the end of the program, then you can simply `mem::forget` the BTreeMap and allow the OS to take over deallocation.

**Comment by the8472** [comment]
Another option is to call `process:exit` which will skip drops.

**Comment by VorpalBlade** [comment]
If it is not the end of the program there are also some options:
* You could offload the destruction on a background thread.
* You could use a map that is allocated from a memory pool that can be dropped all at once (unfortunately not on stable as the allocator API seems stuck on nightly with no end in sight, though there are alternative implementations in crates, that I haven't tested).

Otherwise I believe it is hard to fix since a btreemap contains lots of small allocations, and freeing them individually takes time.

**Comment by the8472** [comment]
> Otherwise I believe it is hard to fix since a btreemap contains lots of small allocations, and freeing them individually takes time.

Switching to a different allocator might help if its internal bookkeeping is faster and it's not already dominated by the function call overhead. Jemalloc has an option to perform cleanups in a background thread for example.

If that doesn't help we'd need a new deallocator API that does batch deallocations.

**Comment by VorpalBlade** [comment]
> If that doesn't help we'd need a new deallocator API that does batch deallocations.

Wouldn't the allocator and/or storage APIs help with this? If you allocate from a bump allocator or set of slabs you can just deallocate the whole thing at once. You can't do batch deallocation faster in the general case I think, since the allocations might not be together but be fragmented all over the address space together with other allocations.

**Comment by the8472** [comment]
Yes, custom allocators could help with that. But I mean even for the system allocator it might be possible to amortize function call/bookkeeping overheads with a batch API. Not that any allocator currently offers those.

> EDIT: This is assuming you don't want to run any custom Drop for your tree nodes of course. If you have that, there isn't much you can do.

`Drop` could also be extended with a `drop_many::<const N>([&mut; N])`, which then could internally make use of a batch deallocation API.
